src/interaction/sessionrecording.cpp

```diff
@@ -0,0 +1,1207 @@
+            else if (entryType.compare("time") == 0) {
+                playbackTimeChange();
+            }
+            else if (entryType.compare("script") == 0) {
+                playbackScript();
+            }
+            else {
+                LERROR(fmt::format("Unknown frame type {} @ line {} of playback file {}",
+                    entryType, _playbackLineNum, _playbackFilename.c_str()));
+                parsingErrorsFound = true;
+                break;
+            }
+        }
+        LINFO(fmt::format("Finished parsing {} entries from playback file {}",
+            _playbackLineNum, _playbackFilename.c_str()));
+    }
+
+    ghoul_assert(
+        std::is_sorted(
+            _timeline.begin(),
+            _timeline.end(),
+            [](const timelineEntry& lhs, const timelineEntry& rhs) { return lhs.timestamp < rhs.timestamp; }),
+        ""
+        );
+
+    return !parsingErrorsFound;
+}
+
+double SessionRecording::getAppropriateTimestamp(double timeOs,
+                                                 double timeRec,
+                                                 double timeSim)
+{
+    if (_playbackTimeReferenceMode == KeyframeTimeRef::relative_recordedStart)
+        return timeRec;
+    else if (_playbackTimeReferenceMode == KeyframeTimeRef::absolute_simTimeJ2000)
+        return timeSim;
+    else
+        return timeOs;
+}
+
+double SessionRecording::getEquivalentSimulationTime(double timeOs,
+                                                     double timeRec,
+                                                     double timeSim)
+{
+    if (_playbackTimeReferenceMode == KeyframeTimeRef::relative_recordedStart)
+        return _timestampPlaybackStarted_simulation + timeRec;
+    else if (_playbackTimeReferenceMode == KeyframeTimeRef::relative_applicationStart)
+        return _timestampApplicationStarted_simulation + timeOs;
+    else
+        return timeSim;
+}
+
+double SessionRecording::getEquivalentApplicationTime(double timeOs,
+                                                      double timeRec,
+                                                      double timeSim)
+{
+    if (_playbackTimeReferenceMode == KeyframeTimeRef::relative_recordedStart)
+        return _timestampPlaybackStarted_application + timeRec;
+    else if (_playbackTimeReferenceMode == KeyframeTimeRef::absolute_simTimeJ2000)
+        return timeSim - _timestampApplicationStarted_simulation;
+    else
+        return timeOs;
+}
+
+double SessionRecording::currentTime() const {
+    if (_playbackTimeReferenceMode == KeyframeTimeRef::relative_recordedStart)
+        return (global::windowDelegate.applicationTime() - _timestampPlaybackStarted_application);
+    else if (_playbackTimeReferenceMode == KeyframeTimeRef::absolute_simTimeJ2000)
+        return global::timeManager.time().j2000Seconds();
+    else
+        return global::windowDelegate.applicationTime();
+}
+    
+void SessionRecording::playbackCamera() {
+    double timeRef;
```

Comment:
This is not your grandpa's C anymore ;)  No need to define all variables at the top of the function

Reply:
They are defined at top because they are used in either binary or ascii cases